src/oic/oauth2/provider.py

```diff
@@ -317,6 +363,145 @@ def _verify_redirect_uri(self, areq):
+        """
+        if pcr_class is not None:
+            warnings.warn(
+                "`pcr_class` is deprecated, please use `message_factory`.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        else:
+            pcr_class = self.server.message_factory.get_response_type(
+                "configuration_endpoint"
+            )
+
+        _provider_info = pcr_class(**self.default_capabilities)
+        _provider_info["scopes_supported"] = self.scopes
+
+        sign_algs = list(jws.SIGNER_ALGS.keys())
+        sign_algs.remove("none")
+        sign_algs = sorted(sign_algs, key=cmp_to_key(sort_sign_alg))
+
+        _pat1 = "{}_endpoint_auth_signing_alg_values_supported"
+        _pat2 = "{}_endpoint_auth_methods_supported"
+        for typ in ["token", "revocation", "introspection"]:
+            _provider_info[_pat1.format(typ)] = sign_algs
+            _provider_info[_pat2.format(typ)] = AUTH_METHODS_SUPPORTED
+
+        if provider_config:
+            _provider_info.update(provider_config)
+
+        return _provider_info
+
+    def create_providerinfo(self, pcr_class=None, setup=None):
+        """
+        Dynamically create the provider info response.
+
+        :param pcr_class:
+        :param setup:
+        :return:
+        """
+        if pcr_class is not None:
+            warnings.warn(
+                "Passing `pcr_class` is deprecated. Please use `message_factory.",
+                DeprecationWarning,
+                stacklevel=2,
+            )
+        else:
+            pcr_class = self.server.message_factory.get_response_type(
+                "configuration_endpoint"
+            )
+        _provider_info = copy.deepcopy(self.capabilities.to_dict())
+
+        if self.jwks_uri and self.keyjar:
+            _provider_info["jwks_uri"] = self.jwks_uri
+
+        for endp in self.endp:
+            if not self.baseurl.endswith("/"):
+                baseurl = self.baseurl + "/"
+            else:
+                baseurl = self.baseurl
+            _provider_info["{}_endpoint".format(endp.etype)] = urljoin(
+                baseurl, endp.url
+            )
+
+        if setup and isinstance(setup, dict):
+            for key in pcr_class.c_param.keys():
+                if key in setup:
+                    _provider_info[key] = setup[key]
+
+        _provider_info["issuer"] = self.name
+        _provider_info["version"] = "3.0"
+
+        return pcr_class(**_provider_info)
+
+    def providerinfo_endpoint(self, handle="", **kwargs):
+        _log_info = logger.info
+
+        _log_info("@providerinfo_endpoint")
+        try:
+            _response = self.create_providerinfo()
+            msg = "provider_info_response: {}"
+            _log_info(msg.format(sanitize(_response.to_dict())))
+            if self.events:
+                self.events.store("Protocol response", _response)
+
+            headers = [("Cache-Control", "no-store"), ("x-ffo", "bar")]
```

Comment:
Any idea what the x-ffo header is for?

Reply:
Hm, could stand for `X-Forwarded-For`? But the setting (and the header itself) are nonsense.

I am for removing it as it is not in a list of standard and common non-standard headers.